[MLIR][XeGPU] Add unroll patterns and blocking pass for XeGPU (1/N)

chencha3 opened the pull request:
Similar to vector ops, XeGPU ops need to be unrolled into smaller shapes such that they can be dispatched into a hardware instruction. This PR marks the initial phase of a series dedicated to incorporating unroll patterns for XeGPU operations. In this installment, we introduce patterns for the following operations:
1. createNd
2. updateNd
3. prefetchNd
4. loadNd
5. storeNd
6. dpas

chencha3 committed:
init

chencha3 committed:
Merge branch 'main' into xegpu_unroll_patterns

chencha3 committed:
Merge branch 'main' into xegpu_unroll_patterns

chencha3 committed:
add patterns for createNdOp and StoreNdOp

chencha3 committed:
refine nativeShapeFn

chencha3 committed:
refine verifier for TensorDescType

chencha3 committed:
add loadNd pattern

chencha3 committed:
add test pass

chencha3 committed:
format code

chencha3 committed:
add unit test

chencha3 committed:
clean up

chencha3 committed:
stage

chencha3 committed:
Merge branch 'main' into xegpu_unroll_patterns

chencha3 committed:
add dpas pattern and unit test

chencha3 committed:
refactor

chencha3 committed:
fix format

chencha3 committed:
fix format

chencha3 committed:
refine

chencha3 committed:
refine

chencha3 committed:
cleanup and add patterns for rest nd ops

chencha3 committed:
fix format

chencha3 committed:
cleanup

Garra1980 commented:
> This PR marks the initial phase of a series dedicated to incorporating unroll patterns for XeGPU operations. 

Can you please add some justification/explanation regarding those unroll patterns

adam-smnk commented on mlir/include/mlir/Dialect/XeGPU/Transforms/Transforms.h:
```diff
@@ -12,11 +12,18 @@
 void populateXeGPUFoldAliasOpsPatterns(RewritePatternSet &patterns);
 
+void populateXeGPUUnrollPatterns(RewritePatternSet &patterns,
+                                 const vector::UnrollVectorOptions &options);
```
It's better to decouple from the generic vector infrastructure and create XeGPU specific options (even if they end up being the same).

chencha3 replied:
> It's better to decouple from the generic vector infrastructure and create XeGPU specific options (even if they end up being the same).

Even these patterns are to be used together with vector unroll patterns in the same pass?

chencha3 replied:
@adam-smnk decoupled it as you suggested.

adam-smnk reviewed (commented):
nit: missing end new lines

chencha3 committed:
add UnrollOption

chencha3 committed:
fix the format

chencha3 committed:
add comments

chencha3 committed:
add brief description

charithaintc commented on mlir/test/Dialect/XeGPU/xegpu-unroll-patterns.mlir:
```diff
@@ -0,0 +1,98 @@
+  //CHECK-COUNT-6: [[tdesc:%.+]] = xegpu.create_nd_tdesc [[arg0]][{{.*}}] : memref<24x32xf32> -> !xegpu.tensor_desc<8x16xf32>
+  //CHECK-COUNT-6: [[data:%.+]] = xegpu.load_nd {{.*}}  : !xegpu.tensor_desc<8x16xf32> -> vector<8x16xf32>
+  //CHECK-COUNT-6: [[insert:%.+]] = vector.insert_strided_slice {{.*}} : vector<8x16xf32> into vector<24x32xf32>
+  //CHECK: [[add:%.+]] = arith.addf {{.*}} : vector<24x32xf32>
```
in this case, what happens with the addi, doesn't it need unrolling as well?

chencha3 replied:
It is going to be unrolled using Vector Unroll patterns. So it is not tested here.

charithaintc commented on mlir/test/lib/Dialect/XeGPU/TestXeGPUTransforms.cpp:
```diff
@@ -0,0 +1,96 @@
+
+namespace {
+
+struct TestXeGPUUnrollingPatterns
```
should we also move SIMT distribution testing here as well? I am just wondering.

chencha3 replied:
Here is to test patterns by writing a test pass. I think your SIMT distribution is already implemented as a pass. Maybe not needed.

charithaintc reviewed (commented):
overall LGTM. I will have another look after addressing the comments.

chencha3 commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,480 @@
+      : OpRewritePattern<SourceOp>(context, benefit), options(options) {}
+
+protected:
+  std::optional<SmallVector<int64_t>> getTargetShape(Operation *op) const {
```
I intentionally to reuse the name from Vector Unroll here to keep it consistent. They have similar meanings, just in different context. nativeShape is more in the context of implementing Unroll Option, referring to compute a shape for an operation. targetShape is more in the context of implementing patterns, referring to the shape for the op to unroll to.

charithaintc replied:
ok sounds good.

chencha3 commented on mlir/include/mlir/Dialect/XeGPU/Transforms/Transforms.h:
```diff
@@ -14,9 +14,36 @@ class RewritePatternSet;
+
+  using NativeShapeFnType =
+      std::function<std::optional<SmallVector<int64_t>>(Operation *op)>;
+  /// Function that returns the shape of the vector to unroll to for a given
```
Any suggestion?

chencha3 replied:
NVM, I asked the help from ChatGPT.

chencha3 commented on mlir/include/mlir/Dialect/XeGPU/IR/XeGPUAttrs.td:
```diff
@@ -274,6 +296,16 @@ def XeGPU_LayoutAttr : XeGPUAttr<"Layout", "layout"> {
+    LayoutAttr dropInstData() {
+      return LayoutAttr::get(getContext(), getSgLayout(), getSgData(), nullptr,
+                             getLaneLayout(), getLaneData(), getOrder());
+    }
```
logically `dropLaneLayoutAndData` is going to return a nullptr. I saw you implemented a dropLayout, I feel that is what you may need.

charithaintc replied:
yeah, ignore this comment. I added it in my PR.

chencha3 commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,480 @@
+      assert(maybeGrids && "Expecting grids to be computed.");
+      auto grids = *maybeGrids;
+      auto numNewOps = computeProduct(grids);
+      if (numNewOps == 1)
```
If the shape is already in the target shape, we don't need to pack it. e.g., given A: 8x16xf16, B: 16x32xf16. we only need to unroll B

chencha3 committed:
address comments

chencha3 committed:
Merge branch 'main' into xegpu_unroll_patterns

chencha3 commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,480 @@
+    if (c)
+      cVals = packWrapper(c, cBlockSize);
+
+    // Vals are empty due to invalid blocking size, or with size 1 due to
+    // the original shape is the same with the blocking size. The op will
+    // be skipped if every operand got an invalid blocking size or the
+    // original shape is the same with the blocking size.
```
Fixed the comment issue. Yes, the blocking size is given by user. If any of the block size is invalid. the unrolling will skip it.

chencha3 commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,480 @@
+    return layout.dropInstData();
+  };
+
+  SmallVector<Type> convertType(ShapedType type,
```
changed it

chencha3 commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,480 @@
+  // emulate the the unpack behavior using insert_strided_slice for VectorType
+  // values and unrealized_conversion_cast for TileType values.
+  Value unpack(ValueRange srcs, Type destTy, llvm::ArrayRef<int64_t> blockSize,
+               Location loc, PatternRewriter &rewriter) const {
```
I simply followed the naming of tensor.pack and tensor.unpack, and earlier proposal of pack/unpack ops from jianhui.

charithaintc replied:
sounds good.

chencha3 committed:
add comments

chencha3 committed:
fix comments

chencha3 commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,480 @@
+  }
+};
+
+struct UnrollDpasOp : public UnrollPattern<xegpu::DpasOp> {
```
Added a general description at declaration of `populateXeGPUUnrollPatterns`

chencha3 committed:
renaming

chencha3 commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,480 @@
+  }
+
+  // copy the layout attribte and drops the inst_data field.
+  xegpu::LayoutAttr getLaneLayoutAttr(Attribute attr) const {
```
fixed

charithaintc commented on mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td:
```diff
@@ -38,6 +38,8 @@ def XeGPU_Dialect : Dialect {
+      static constexpr const char *resultLayoutNamePrefix = "layout_result_";
       /// Checks if the given shape can be evenly distributed based on the layout
       /// and data factors provided by the LayoutAttr.
       static bool isEvenlyDistributable(llvm::ArrayRef<int64_t> shape, xegpu::LayoutAttr attr);
```
I see.

Garra1980 commented on mlir/include/mlir/Dialect/XeGPU/Transforms/Transforms.h:
```diff
@@ -14,11 +14,53 @@ class RewritePatternSet;
 
+/// Collect a set of pattern to unroll xegpu operations to a smaller shapes.
+/// Users can control whether an operation to be unrolled or not, as well as
+/// the its target shape via `options` structure. (via setting filterConstraint
```
'the' is redundant

chencha3 replied:
Fixed, thanks

Garra1980 commented on mlir/include/mlir/Dialect/XeGPU/Transforms/Transforms.h:
```diff
@@ -14,11 +14,53 @@ class RewritePatternSet;
 /// Appends patterns for XeGPU SIMT distribution into `patterns`.
 void populateXeGPUSubgroupDistributePatterns(RewritePatternSet &patterns);
 
+/// Collect a set of pattern to unroll xegpu operations to a smaller shapes.
```
pattern->patterns

Garra1980 commented on mlir/include/mlir/Dialect/XeGPU/Transforms/Transforms.h:
```diff
@@ -14,11 +14,53 @@ class RewritePatternSet;
+  /// Callback function that indicates whether vector unrolling should be
+  /// attempted on the operation.
+  FilterConstraintFnType filterConstraint = nullptr;
+  UnrollOptions &setFilterConstraint(FilterConstraintFnType constraint) {
```
Is there an example/test that demonstrates usage of this option?

Garra1980 commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,497 @@
+                                     ArrayRef<int64_t> blockSize) const {
+    auto elemTy = type.getElementType();
+    Type newTy;
+    // TensorDescType needs to drop the inst_data field in the layout attribute
```
Why?

chencha3 replied:
After unroll, the data shape is supposed to instruction size, so the inst_data is meaningless.

chencha3 replied:
I moved this logic out.

Garra1980 commented on mlir/include/mlir/Dialect/XeGPU/Transforms/Transforms.h:
```diff
@@ -14,11 +14,53 @@ class RewritePatternSet;
+///   BuildinUnrealizedCastop are created to break-up the TensorDesc operands.
+///   3. the original op is cloned `numUnrolledInstances` times, once for each
+///   result.
+///   4. InsertStridedSlice are inserted for VectorType result, and
```
in 2. and 4. you mean functionality of functions named 'pack' and 'unpack'?

chencha3 replied:
Yes

chencha3 committed:
generalize pack, unpack, createaNdOp for supporting 1D cases

chencha3 committed:
refine

chencha3 committed:
add 1D unit tests

chencha3 reviewed (commented):
> Is it documented somewhere? 

Good point. This reminds me this logic shouldn't belong here.

fschlimb commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,466 @@
+    return nativeShape;
+  }
+
+  // std::optional<SmallVector<int64_t>>
+  // computeGrids(llvm::ArrayRef<int64_t> shape,
+  //              llvm::ArrayRef<int64_t> subShape) const {
+  //   // if the shape == subshape, we don't need to unroll.
+  //   if (shape == subShape) {
+  //     LDBG("shape == subshape, no unroll");
+  //     return std::nullopt;
+  //   }
+  //   return computeShapeRatio(shape, subShape);
+  // }
```
remove?

chencha3 replied:
Thanks! removed.

fschlimb commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,466 @@
+  //   return computeShapeRatio(shape, subShape);
+  // }
+
+  // copy the layout attribte and drops the inst_data field.
```
typo

chencha3 replied:
Thanks! Fixed

fschlimb commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,466 @@
+    auto b = op.getRhs();
+    auto c = op.getAcc();
+
+    auto aShape = a.getType().getShape();
+    auto bShape = b.getType().getShape();
```
Can these be dynamic shapes?

chencha3 replied:
No. the dpas definition doesn't support dynamic shape.

fschlimb commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,466 @@
+                                llvm::ArrayRef<int64_t> blockSize, Location loc,
+                                PatternRewriter &rewriter) const {
+    if (auto vecTy = dyn_cast<VectorType>(src.getType())) {
+      assert(vecTy.getRank() == (int64_t)blockSize.size() &&
```
LLVM coding standard wants c++ casts: https://llvm.org/docs/CodingStandards.html#prefer-c-style-casts

chencha3 replied:
Thanks! Fixed. It seems function-like casts are fine for Integers?

fschlimb replied:
Yes, but these are not function-like casts. If you prefer those, it should be `int64_t(blockSize.size())`

fschlimb commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,466 @@
+    auto nIters = bShape[1] / N;
+
+    SmallVector<Value> newOps;
+    for (int64_t i = 0; i < mIters; i++) {
+      for (int64_t j = 0; j < nIters; j++) {
+        Value tmpC;
+        if (c)
+          tmpC = cVals[i * nIters + j]; // init with acc
+        for (int64_t k = 0; k < kIters; k++) {
```
LLVM coding standards want preincrements (++i): https://llvm.org/docs/CodingStandards.html#prefer-preincrement

chencha3 replied:
Thanks! Fixed.

fschlimb commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,466 @@
+    auto targetShape = *maybeTargetShape;
+
+    auto convertedTdescTypes = getUnrolledTypes(tdescTy, targetShape);
+    auto convertedTdesc =
```
Use explicit type instead of auto: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

fschlimb replied:
Many place in this code might be more readable with explicit types.

chencha3 replied:
Thanks. Fixed.

fschlimb commented on mlir/test/lib/Dialect/XeGPU/TestXeGPUTransforms.cpp:
```diff
@@ -0,0 +1,96 @@
+            }
+          }
+
+          if (isa<xegpu::DpasOp>(op)) {
```
no {}

chencha3 replied:
thanks! fixed.

fschlimb commented:
Added a few minor comments. On the monkey-level this looks good to me.
You might want to capitalize the first words in paragraphs and enums/bullet lists within comments.

chencha3 committed:
switch to explicit types

chencha3 committed:
clean up

chencha3 commented:
> Added a few minor comments. On the monkey-level this looks good to me. You might want to capitalize the first words in paragraphs and enums/bullet lists within comments.

Thanks @fschlimb, I made the changes according to your feedback. I hope I addressed all of your concerns.

charithaintc reviewed (approved):
LGTM

adam-smnk reviewed (commented):
It'll be interesting to later see if we could generalize and reuse vector unrolling to achieve the same. For now, I think it's a good addition to xegpu infrastructure and we'll see in practice how it holds up.

I take it depends on #138701?

adam-smnk commented on mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:
```diff
@@ -141,46 +142,24 @@ void CreateNdDescOp::build(OpBuilder &builder, OperationState &state,
-  llvm::SmallVector<int64_t> staticOffsets;
-  llvm::SmallVector<int64_t> staticShape;
-  llvm::SmallVector<int64_t> staticStrides;
+  auto intTy = dyn_cast<IntegerType>(source.getType());
+  auto memrefTy = dyn_cast<MemRefType>(source.getType());
+  assert(intTy || memrefTy && "Source has to be either int or memref.");
```
nit: maybe let's cache type itself, use two `isa<>` here and only `dyn_cast` further below

chencha3 replied:
Thanks! Fixed.

adam-smnk commented on mlir/test/lib/Dialect/XeGPU/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+  MLIRTransforms
+  MLIRGPUDialect
+  MLIRXeGPUTransforms
+)
```
nit: new line

chencha3 replied:
Thanks! fixed

chencha3 committed:
move getUnrolledTypes out

chencha3 commented:
> It'll be interesting to later see if we could generalize and reuse vector unrolling to achieve the same. For now, I think it's a good addition to xegpu infrastructure and we'll see in practice how it holds up.
> 
> I take it depends on #138701?

Yeah. these patterns are supposed to be companions to vector unrolling patterns. They share the same idea, one is for XeGPU ops only, and one is for vector ops. A pass are supposed to use both of them.

chencha3 committed:
addressed comments

chencha3 commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,466 @@
+    if (any_of(ranges, [](auto &v) { return v.size() == 0; }) ||
+        all_of(ranges, [](auto &v) { return v.size() == 1; })) {
+      return failure();
+    }
```
Thanks! Fixed.

chencha3 committed:
address comments

chencha3 committed:
fix format

chencha3 committed:
Merge branch 'main' into xegpu_unroll_patterns

chencha3 committed:
sync

Garra1980 commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,427 @@
+  }
+
+private:
+  const char *const packAttrName = "__xetile_blocking_pack__";
```
xetile->xegpu I guess, here and in  tests

chencha3 replied:
good catch. fixed it

Garra1980 commented on mlir/include/mlir/Dialect/XeGPU/Transforms/Transforms.h:
```diff
@@ -14,11 +14,67 @@ class RewritePatternSet;
+/// provide a way to customize the native shape of the operation.
+struct UnrollOptions {
+  using FilterConstraintFnType = std::function<LogicalResult(Operation *op)>;
+  /// Callback function that indicates whether vector unrolling should be
```
nit: let's place this comment above "using" to have uniform look :)

chencha3 replied:
fixed it

chencha3 committed:
address comments

chencha3 committed:
Merge branch 'main' into xegpu_unroll_patterns

chencha3 commented:
> It'll be interesting to later see if we could generalize and reuse vector unrolling to achieve the same. For now, I think it's a good addition to xegpu infrastructure and we'll see in practice how it holds up.
> 
> I take it depends on #138701?

#138701 has been merged, and this PR is rebased too.

chencha3 commented:
Hi @fschlimb and @adam-smnk, do you have more suggestions?

fschlimb commented:
> Hi @fschlimb and @adam-smnk, do you have more suggestions?

no, LGTM!

chencha3 committed:
update cmake

chencha3 commented:
Hi @adam-smnk, I am going to merge this first. If you have more suggestions, feel free to let me know. Thanks for your help!

mgorny commented:
I'm seeing a regression on 32-bit platforms from this change:

```
FAIL: MLIR :: Dialect/XeGPU/xegpu-unroll-patterns.mlir (1488 of 2894)
******************** TEST 'MLIR :: Dialect/XeGPU/xegpu-unroll-patterns.mlir' FAILED ********************
Exit Code: 2

Command Output (stdout):
--
# RUN: at line 1
/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/mlir-opt --test-xegpu-unrolling-patterns -split-input-file /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir/test/Dialect/XeGPU/xegpu-unroll-patterns.mlir | /usr/lib/llvm/21/bin/FileCheck /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir/test/Dialect/XeGPU/xegpu-unroll-patterns.mlir
# executed command: /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/mlir-opt --test-xegpu-unrolling-patterns -split-input-file /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir/test/Dialect/XeGPU/xegpu-unroll-patterns.mlir
# .---command stderr------------
# | mlir-opt: /usr/lib/llvm/21/include/llvm/ADT/SmallVector.h:291: T& llvm::SmallVectorTemplateCommon<T, <template-parameter-1-2> >::operator[](size_type) [with T = long long int; <template-parameter-1-2> = void; reference = long long int&; size_type = unsigned int]: Assertion `idx < size()' failed.
# | PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
# | Stack dump:
# | 0.  Program arguments: /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/mlir-opt --test-xegpu-unrolling-patterns -split-input-file /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir/test/Dialect/XeGPU/xegpu-unroll-patterns.mlir
# |  #0 0xffffffffe5e86ef0 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/lib/llvm/21/lib/libLLVM.so.21.0gitaaaae996+0xc45ef0)
# |  #1 0xffffffffe5e8747f (/usr/lib/llvm/21/lib/libLLVM.so.21.0gitaaaae996+0xc4647f)
# |  #2 0xffffffffe5e84095 llvm::sys::RunSignalHandlers() (/usr/lib/llvm/21/lib/libLLVM.so.21.0gitaaaae996+0xc43095)
# |  #3 0xffffffffe5e8422b (/usr/lib/llvm/21/lib/libLLVM.so.21.0gitaaaae996+0xc4322b)
# |  #4 0xfffffffff7f8d5a0 (linux-gate.so.1+0x5a0)
# |  #5 0xfffffffff7f8d579 (linux-gate.so.1+0x579)
# |  #6 0xffffffffe4de7d07 (/usr/lib/libc.so.6+0x93d07)
# |  #7 0xffffffffe4d8c581 raise (/usr/lib/libc.so.6+0x38581)
# |  #8 0xffffffffe4d732d8 abort (/usr/lib/libc.so.6+0x1f2d8)
# |  #9 0xffffffffe4d731de (/usr/lib/libc.so.6+0x1f1de)
# | #10 0xffffffffe4d8450b (/usr/lib/libc.so.6+0x3050b)
# | #11 0xfffffffff58af00d mlir::computeSuffixProduct(llvm::ArrayRef<long long>) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x4c3200d)
# | #12 0xfffffffff58b176c mlir::detail::TileOffsetRangeImpl::TileOffsetRangeImpl(llvm::ArrayRef<long long>, llvm::ArrayRef<long long>, llvm::ArrayRef<long long>) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x4c3476c)
# | #13 0xfffffffff2db29fc mlir::StaticTileOffsetRange::StaticTileOffsetRange(llvm::ArrayRef<long long>, llvm::ArrayRef<long long>) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x21359fc)
# | #14 0xfffffffff5b7c52a (anonymous namespace)::UnrollPattern<mlir::xegpu::StoreNdOp>::pack(mlir::Value, mlir::TypeRange, llvm::ArrayRef<long long>, mlir::Location, mlir::PatternRewriter&) const XeGPUUnroll.cpp:0:0
# | #15 0xfffffffff5b7ef4e (anonymous namespace)::UnrollStoreNdOp::matchAndRewrite(mlir::xegpu::StoreNdOp, mlir::PatternRewriter&) const XeGPUUnroll.cpp:0:0
# | #16 0xfffffffff5b7b2ee mlir::detail::OpOrInterfaceRewritePatternBase<mlir::xegpu::StoreNdOp>::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x4efe2ee)
# | #17 0xfffffffff5da80ba mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<llvm::LogicalResult (mlir::Pattern const&)>)::'lambda'()::operator()() const PatternApplicator.cpp:0:0
# | #18 0xfffffffff5da8f4e mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<llvm::LogicalResult (mlir::Pattern const&)>) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x512bf4e)
# | #19 0xfffffffff61ade63 (anonymous namespace)::GreedyPatternRewriteDriver::processWorklist() GreedyPatternRewriteDriver.cpp:0:0
# | #20 0xfffffffff61b029e mlir::applyPatternsGreedily(mlir::Region&, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig, bool*) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x553329e)
# | #21 0x5700059d (anonymous namespace)::TestXeGPUUnrollingPatterns::runOnOperation() TestXeGPUTransforms.cpp:0:0
# | #22 0xfffffffff5d6d371 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x50f0371)
# | #23 0xfffffffff5d6d8b4 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x50f08b4)
# | #24 0xfffffffff5d6dd8d mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::'lambda'(mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo&)::operator()(mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo&) const Pass.cpp:0:0
# | #25 0xfffffffff5d6c3a4 mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/
work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x50ef3a4)
# | #26 0xfffffffff5d6d094 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned in
t) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x50f0094)
# | #27 0xfffffffff5d6d8b4 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, 
bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (/var/tmp/portage/llvm-core/mlir-21.
0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x50f08b4)
# | #28 0xfffffffff5d6ed93 mlir::PassManager::run(mlir::Operation*) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x50f1d93)
# | #29 0xfffffffff613a8e6 performActions(llvm::raw_ostream&, std::shared_ptr<llvm::SourceMgr> const&, mlir::MLIRContext*, mlir::MlirOptMainConfig const&) MlirOptMain.cpp:0:0
# | #30 0xfffffffff613b1d0 llvm::LogicalResult llvm::function_ref<llvm::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>::callback_fn<mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&)::'lambda'(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>(int, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&) MlirOptMain.cpp:0:0
# | #31 0xfffffffff5df3ceb mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<llvm::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, llvm::StringRef, llvm::StringRef)::'lambda'(llvm::StringRef)::operator()(llvm::StringRef) const ToolUtilities.cpp:0:0
# | #32 0xfffffffff5df4425 mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<llvm::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, llvm::StringRef, llvm::StringRef) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x5177425)
# | #33 0xfffffffff6132e39 mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x54b5e39)
# | #34 0xfffffffff613b516 mlir::MlirOptMain(int, char**, llvm::StringRef, llvm::StringRef, mlir::DialectRegistry&) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x54be516)
# | #35 0xfffffffff613ba51 mlir::MlirOptMain(int, char**, llvm::StringRef, mlir::DialectRegistry&) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x54bea51)
# | #36 0x568ff7ac main (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/mlir-opt+0x3237ac)
# | #37 0xffffffffe4d74f83 (/usr/lib/libc.so.6+0x20f83)
# | #38 0xffffffffe4d75048 __libc_start_main (/usr/lib/libc.so.6+0x21048)
# | #39 0x568fff97 _start (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/mlir-opt+0x323f97)
# `-----------------------------
# error: command failed with exit status: -6
# executed command: /usr/lib/llvm/21/bin/FileCheck /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir/test/Dialect/XeGPU/xegpu-unroll-patterns.mlir
# .---command stderr------------
# | FileCheck error: '<stdin>' is empty.
# | FileCheck command line:  /usr/lib/llvm/21/bin/FileCheck /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir/test/Dialect/XeGPU/xegpu-unroll-patterns.mlir
# `-----------------------------
# error: command failed with exit status: 2

--

********************
```

Jianhui-Li commented on mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp:
```diff
@@ -0,0 +1,427 @@
+      auto shape = vecTy.getShape();
+      SmallVector<Value> results;
+      for (SmallVector<int64_t> offsets :
+           StaticTileOffsetRange(shape, blockSize)) {
```
The PR seems not handling xegpu.order attribute at this stage. load_nd on PVC support load with transpose, which takes order as [0,1].  
StaticTileOffsetRange() can take extra loopOrder parameter which can be used to support this feature.

chencha3 commented:
> I'm seeing a regression on 32-bit platforms from this change:
> 
> ```
> FAIL: MLIR :: Dialect/XeGPU/xegpu-unroll-patterns.mlir (1488 of 2894)
> ******************** TEST 'MLIR :: Dialect/XeGPU/xegpu-unroll-patterns.mlir' FAILED ********************
> Exit Code: 2
> 
> Command Output (stdout):
> --
> # RUN: at line 1
> /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/mlir-opt --test-xegpu-unrolling-patterns -split-input-file /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir/test/Dialect/XeGPU/xegpu-unroll-patterns.mlir | /usr/lib/llvm/21/bin/FileCheck /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir/test/Dialect/XeGPU/xegpu-unroll-patterns.mlir
> # executed command: /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/mlir-opt --test-xegpu-unrolling-patterns -split-input-file /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir/test/Dialect/XeGPU/xegpu-unroll-patterns.mlir
> # .---command stderr------------
> # | mlir-opt: /usr/lib/llvm/21/include/llvm/ADT/SmallVector.h:291: T& llvm::SmallVectorTemplateCommon<T, <template-parameter-1-2> >::operator[](size_type) [with T = long long int; <template-parameter-1-2> = void; reference = long long int&; size_type = unsigned int]: Assertion `idx < size()' failed.
> # | PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
> # | Stack dump:
> # | 0.  Program arguments: /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/mlir-opt --test-xegpu-unrolling-patterns -split-input-file /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir/test/Dialect/XeGPU/xegpu-unroll-patterns.mlir
> # |  #0 0xffffffffe5e86ef0 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/lib/llvm/21/lib/libLLVM.so.21.0gitaaaae996+0xc45ef0)
> # |  #1 0xffffffffe5e8747f (/usr/lib/llvm/21/lib/libLLVM.so.21.0gitaaaae996+0xc4647f)
> # |  #2 0xffffffffe5e84095 llvm::sys::RunSignalHandlers() (/usr/lib/llvm/21/lib/libLLVM.so.21.0gitaaaae996+0xc43095)
> # |  #3 0xffffffffe5e8422b (/usr/lib/llvm/21/lib/libLLVM.so.21.0gitaaaae996+0xc4322b)
> # |  #4 0xfffffffff7f8d5a0 (linux-gate.so.1+0x5a0)
> # |  #5 0xfffffffff7f8d579 (linux-gate.so.1+0x579)
> # |  #6 0xffffffffe4de7d07 (/usr/lib/libc.so.6+0x93d07)
> # |  #7 0xffffffffe4d8c581 raise (/usr/lib/libc.so.6+0x38581)
> # |  #8 0xffffffffe4d732d8 abort (/usr/lib/libc.so.6+0x1f2d8)
> # |  #9 0xffffffffe4d731de (/usr/lib/libc.so.6+0x1f1de)
> # | #10 0xffffffffe4d8450b (/usr/lib/libc.so.6+0x3050b)
> # | #11 0xfffffffff58af00d mlir::computeSuffixProduct(llvm::ArrayRef<long long>) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x4c3200d)
> # | #12 0xfffffffff58b176c mlir::detail::TileOffsetRangeImpl::TileOffsetRangeImpl(llvm::ArrayRef<long long>, llvm::ArrayRef<long long>, llvm::ArrayRef<long long>) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x4c3476c)
> # | #13 0xfffffffff2db29fc mlir::StaticTileOffsetRange::StaticTileOffsetRange(llvm::ArrayRef<long long>, llvm::ArrayRef<long long>) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x21359fc)
> # | #14 0xfffffffff5b7c52a (anonymous namespace)::UnrollPattern<mlir::xegpu::StoreNdOp>::pack(mlir::Value, mlir::TypeRange, llvm::ArrayRef<long long>, mlir::Location, mlir::PatternRewriter&) const XeGPUUnroll.cpp:0:0
> # | #15 0xfffffffff5b7ef4e (anonymous namespace)::UnrollStoreNdOp::matchAndRewrite(mlir::xegpu::StoreNdOp, mlir::PatternRewriter&) const XeGPUUnroll.cpp:0:0
> # | #16 0xfffffffff5b7b2ee mlir::detail::OpOrInterfaceRewritePatternBase<mlir::xegpu::StoreNdOp>::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x4efe2ee)
> # | #17 0xfffffffff5da80ba mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<llvm::LogicalResult (mlir::Pattern const&)>)::'lambda'()::operator()() const PatternApplicator.cpp:0:0
> # | #18 0xfffffffff5da8f4e mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<llvm::LogicalResult (mlir::Pattern const&)>) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x512bf4e)
> # | #19 0xfffffffff61ade63 (anonymous namespace)::GreedyPatternRewriteDriver::processWorklist() GreedyPatternRewriteDriver.cpp:0:0
> # | #20 0xfffffffff61b029e mlir::applyPatternsGreedily(mlir::Region&, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig, bool*) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x553329e)
> # | #21 0x5700059d (anonymous namespace)::TestXeGPUUnrollingPatterns::runOnOperation() TestXeGPUTransforms.cpp:0:0
> # | #22 0xfffffffff5d6d371 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x50f0371)
> # | #23 0xfffffffff5d6d8b4 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x50f08b4)
> # | #24 0xfffffffff5d6dd8d mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::'lambda'(mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo&)::operator()(mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo&) const Pass.cpp:0:0
> # | #25 0xfffffffff5d6c3a4 mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/
> work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x50ef3a4)
> # | #26 0xfffffffff5d6d094 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned in
> t) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x50f0094)
> # | #27 0xfffffffff5d6d8b4 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, 
> bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (/var/tmp/portage/llvm-core/mlir-21.
> 0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x50f08b4)
> # | #28 0xfffffffff5d6ed93 mlir::PassManager::run(mlir::Operation*) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x50f1d93)
> # | #29 0xfffffffff613a8e6 performActions(llvm::raw_ostream&, std::shared_ptr<llvm::SourceMgr> const&, mlir::MLIRContext*, mlir::MlirOptMainConfig const&) MlirOptMain.cpp:0:0
> # | #30 0xfffffffff613b1d0 llvm::LogicalResult llvm::function_ref<llvm::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>::callback_fn<mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&)::'lambda'(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>(int, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&) MlirOptMain.cpp:0:0
> # | #31 0xfffffffff5df3ceb mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<llvm::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, llvm::StringRef, llvm::StringRef)::'lambda'(llvm::StringRef)::operator()(llvm::StringRef) const ToolUtilities.cpp:0:0
> # | #32 0xfffffffff5df4425 mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<llvm::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, llvm::StringRef, llvm::StringRef) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x5177425)
> # | #33 0xfffffffff6132e39 mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x54b5e39)
> # | #34 0xfffffffff613b516 mlir::MlirOptMain(int, char**, llvm::StringRef, llvm::StringRef, mlir::DialectRegistry&) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x54be516)
> # | #35 0xfffffffff613ba51 mlir::MlirOptMain(int, char**, llvm::StringRef, mlir::DialectRegistry&) (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/../lib/libMLIR.so.21.0gitaaaae996+0x54bea51)
> # | #36 0x568ff7ac main (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/mlir-opt+0x3237ac)
> # | #37 0xffffffffe4d74f83 (/usr/lib/libc.so.6+0x20f83)
> # | #38 0xffffffffe4d75048 __libc_start_main (/usr/lib/libc.so.6+0x21048)
> # | #39 0x568fff97 _start (/var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir_build-abi_x86_32.x86/bin/mlir-opt+0x323f97)
> # `-----------------------------
> # error: command failed with exit status: -6
> # executed command: /usr/lib/llvm/21/bin/FileCheck /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir/test/Dialect/XeGPU/xegpu-unroll-patterns.mlir
> # .---command stderr------------
> # | FileCheck error: '<stdin>' is empty.
> # | FileCheck command line:  /usr/lib/llvm/21/bin/FileCheck /var/tmp/portage/llvm-core/mlir-21.0.0.9999/work/mlir/test/Dialect/XeGPU/xegpu-unroll-patterns.mlir
> # `-----------------------------
> # error: command failed with exit status: 2
> 
> --
> 
> ********************
> ```

Hi @mgorny, it is fixed here: https://github.com/llvm/llvm-project/pull/140567

mgorny commented:
A relative simple reproducer would be:

```
CC=i686-pc-linux-gnu-gcc CXX=i686-pc-linux-gnu-g++ cmake ../llvm -G Ninja -DLLVM_CCACHE_BUILD=ON -DLLVM_ENABLE_PROJECTS='llvm;mlir' -DCMAKE_BUILD_TYPE=MinSizeRel
ninja
ninja check-mlir-dialect-xegpu
```

For us, `i686-pc-linux-gnu-gcc` is a trivial wrapper:

```
#!/bin/sh
exec x86_64-pc-linux-gnu-gcc -m32 -mfpmath=sse "${@}"
```

Though I suppose `i686-pc-linux-gnu-clang` symlink would work too.
